File: banco_problemas_aplicados/nivel_basico/soluciones/java/Problema04.java
/*
 * Problema 04 - Codigo de Error en Buffer
 * int buscaElemento(int[] arr, int x, int tam)  -> posicion de x, o -1
 */
import java.util.Scanner;

public class Problema04 {

    static int buscaElemento(int[] arr, int x, int tam) {
        for (int i = 0; i < tam; i++) {
            if (arr[i] == x) {
                return i;
            }
        }
        return -1;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.print("Ingrese el tamano del buffer: ");
        int tam = sc.nextInt();

        int[] arr = new int[tam];
        for (int i = 0; i < tam; i++) {
            System.out.print("Estado " + i + ": ");
            arr[i] = sc.nextInt();
        }

        System.out.print("Ingrese el codigo de falla a buscar: ");
        int x = sc.nextInt();

        int pos = buscaElemento(arr, x, tam);
        if (pos < 0) {
            System.out.println("El codigo " + x + " no aparece en el buffer.");
        } else {
            System.out.println("El codigo " + x + " esta en la posicion: " + pos);
        }
        sc.close();
    }
}

File: banco_problemas_aplicados/nivel_basico/soluciones/c/problema04.c
/*
 * Problema 04 - Codigo de Error en Buffer
 * int buscaElemento(int Arr[], int x, int tam)
 * Contexto: registro de estados de microcontrolador. Devuelve la posicion del codigo x, o -1.
 */
#include <stdio.h>

int buscaElemento(int arr[], int x, int tam) {
    for (int i = 0; i < tam; i++) {
        if (arr[i] == x) {
            return i;                /* encontrado: se devuelve el indice de inmediato */
        }
    }
    return -1;                       /* no esta en el buffer */
}

int main(void) {
    int tam, x;
    printf("Ingrese el tamano del buffer: ");
    scanf("%d", &tam);

    int arr[tam];
    for (int i = 0; i < tam; i++) {
        printf("Estado %d: ", i);
        scanf("%d", &arr[i]);
    }

    printf("Ingrese el codigo de falla a buscar: ");
    scanf("%d", &x);

    int pos = buscaElemento(arr, x, tam);
    if (pos < 0) {
        printf("El codigo %d no aparece en el buffer.\n", x);
    } else {
        printf("El codigo %d esta en la posicion: %d\n", x, pos);
    }
    return 0;
}
